ARM64: add support for TI AM62X

gramsay0 opened the pull request:
This PR adds support for:

1. TI AM62x Cortex-A53 SOC
2. PHYTEC phyCORE-AM62x board

gramsay0 committed:
dts: arm64: Add TI AM6234 A53 device tree

Some register addresses change within the AM6X family.
AM62X is common to AM623X/AM625X processors.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
boards: arm64: Add PHYTEC phyCORE-AM62x A53 board support

PHYTEC phyCORE-AM62x is board is based on TI Sitara applications
processor, composed of a quad Cortex®-A53 cluster and a single
Cortex®-M4 core.

Zephyr OS is ported to run on the Cortex®-A53 core.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
drivers: serial: Name the NS16550 variant Kconfig choice

Naming this choice allows setting a default value in defconfig.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
soc: arm64: Add TI AM6234 SoC A53 support

Add minimal SoC support for the TI AM6X series Cortex-A53 Core.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

DhruvaG2000 commented on boards/arm64/phycore_am62x_a53/doc/index.rst:
```diff
@@ -0,0 +1,117 @@
+
+Programming
+***********
+
```
We can also probably add a build instruction somewhere,

``west build -p -b phycore_am62x_a53 samples/hello_world``

jadonk commented:
This is awesome. Do you have any plans to remove the dependency on u-boot?

DhruvaG2000 commented:
![image](https://user-images.githubusercontent.com/41233856/215951194-fcc4f6ea-bdf2-4bc8-9a44-9a44cb78837c.png)

Attached screenshot of Zephyr SHELL Example up on TI AM625 SK EVM as well (running on A53 core).
Great Work @gramsay0 !

SgrrZhf commented on soc/arm64/ti_sitara/am6x/mmu_regions.c:
```diff
@@ -0,0 +1,65 @@
+			      DT_REG_SIZE_BY_IDX(DT_NODELABEL(gic), 1),
+			      MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("UART0",
```
Do we need to add `UART` items into `mmu_regions[]`?
The `uart_ns16550` driver can map the mmio address by itself.
https://github.com/zephyrproject-rtos/zephyr/blob/63850804f167353df27b6b33d05f8f3ab642aae4/drivers/serial/uart_ns16550.c#L350

gramsay0 replied:
I wasn't aware of that, I will remove it (this actually explains a lot...)

DhruvaG2000 commented on boards/arm64/phycore_am62x_a53/doc/index.rst:
```diff
@@ -0,0 +1,117 @@
+Install U-Boot dependencies:
+https://u-boot.readthedocs.io/en/latest/build/gcc.html#dependencies
+
+Clone the PHYTEC TI fork of U-Boot:
```
Is this really needed? Can you check with upstream UBoot once? AM62 has fairly good upstream uboot support.

gramsay0 replied:
Only these PHYTEC boards are supported in upstream U-Boot:

- phycore-am335x-r2-regor_defconfig
- phycore-am335x-r2-wega_defconfig
- phycore-imx8mm_defconfig
- phycore-imx8mp_defconfig
- phycore-rk3288_defconfig
- phycore_pcl063_defconfig
- phycore_pcl063_ull_defconfig 

But feel free to add TI AM62X EVM boards after this is merged

dnltz replied:
@gramsay0 - Thanks for this PR! We will send our u-boot upstream soon because we internally focus on the kernel first.

I will send a patch and add the cache cmd to our u-boot.

gramsay0 replied:
@dnltz thanks! We can update the docs once that's merged.
`CONFIG_CMD_MP` would also be useful for testing SMP

gramsay0 commented:
> This is awesome. Do you have any plans to remove the dependency on u-boot?

Thanks! Not currently

gramsay0 committed:
scripts: ci: check_compliance: Ignore undefined CONFIG_CMD_CACHE

CONFIG_CMD_CACHE is a U-Boot config, mentioned in docs.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

carlocaione commented:
> Do you want to add yourself under collaborators for AM62 under https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml#L2053 ?

I think we should be a bit more careful with adding as maintainer a first time contributor. No objection as collaborator though.

stephanosio commented:
> Do you want to ad yourself as maintainer for AM62 under https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml#L2053 ?

At this point, the "TI Platforms" area should be split into two different areas: "TI SimpleLink Platforms" and "TI Sitara Platforms," since they are two completely different platforms.

vaishnavachath reviewed (dismissed):
@gramsay0 Great work! :confetti_ball:

gramsay0 committed:
MAINTAINERS: Split TI platforms into SimpleLink and Sitara

These platforms are significantly different.
Add vaishnavachath as a TI Sitara Platforms maintainer.
Add gramsay0 as a TI Sitara Platforms collaborator.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 commented:
I'm not sure why changing MAINTAINERS.yml would break this test:

> tests/lib/mpsc_pbuf/lib.mpsc_pbuf_concurrent (qemu_x86_64:tests/lib/mpsc_pbuf/lib.mpsc_pbuf_concurrent) failed

carlescufi commented:
@gramsay0 can you please rebase this?

dnltz commented on boards/arm64/phycore_am62x_a53/doc/index.rst:
```diff
@@ -0,0 +1,129 @@
+Overview
+********
+
+PHYTEC phyCORE-AM62x is board is based on TI Sitara applications
```
```suggestion
PHYTEC phyCORE-AM62x board is based on TI Sitara applications
```

dnltz commented on boards/arm64/phycore_am62x_a53/doc/index.rst:
```diff
@@ -0,0 +1,129 @@
+Building
+********
+
+You can build application in the usual way. Refer to
```
I'm not a native speaker but feels like there is an article missing.
```suggestion
You can build an application in the usual way. Refer to
```

dnltz reviewed (commented):
I only found two issues in the documentation. Everything else looks good to me. Thanks!

carlescufi commented:
@gramsay0 please note that Zephyr is currently in feature freeze until 3.3.0 is released. See schedule here:
https://github.com/zephyrproject-rtos/zephyr/wiki/Release-Management

vaishnavachath commented on MAINTAINERS.yml:
```diff
@@ -2831,9 +2844,9 @@ x86 arch:
     - nashif
     - galak
+    - stephanosio
   files:
```
@gramsay0 nitpick, I think you should rebase to not change the order of maintainers here, see https://github.com/zephyrproject-rtos/zephyr/commit/f05229c6bc7fed9cdbf670f9e16a94e98f945543

gramsay0 replied:
Fixed

stephanosio commented on MAINTAINERS.yml:
```diff
@@ -2831,9 +2844,9 @@ x86 arch:
 Continuous Integration:
   status: maintained
   maintainers:
-    - stephanosio
     - nashif
     - galak
+    - stephanosio
```
Spurious change

gramsay0 commented:
Can we merge this now that 3.3 is released?
It looks like some jobs were cancelled during the last build, is it possible to rebuild without rejecting reviews?
